hera_pspec/pspecdata.py

```diff
@@ -606,8 +630,79 @@ def p_hat(self, M, q):
         """
         return np.dot(M, q)
 
+    def scalar(self, taper='none', little_h=True, num_steps=10000):
+        """
+        Computes the scalar function to convert a power spectrum estimate
+        in "telescope units" to cosmological units
+
+        See arxiv:1304.4991 and HERA memo #27 for details.
+
+        Currently this is only for Stokes I.
+
+        Parameters
+        ----------
+        taper : str, optional
+                Whether a tapering function (e.g. Blackman-Harris) is being
+                used in the power spectrum estimation.
+                Default 
+
+        little_h : boolean, options
+                Whether to have cosmological length units be h^-1 Mpc or Mpc
+                Default: h^-1 Mpc
+
+        Returns
+        -------
+        scalar: float
+                [\int dnu (\Omega_PP / \Omega_P^2) ( B_PP / B_P^2 ) / (X^2 Y)]^-1
+                in h^-3 Mpc^3 or Mpc^3.
+        """
+
+        # Get \Omega_p = \int dOmega beam and \Omega_pp = \int dOmega beam**2
+        power_beam = np.zeros((self.beam_model_nfreqs,self.beam_model_npix))
+        power_beam_Cl = []
+        power_beam_int = np.zeros(self.beam_model_nfreqs)
+        for i in range(self.beam_model_nfreqs):
+            # FIXME: Should the line below (the sum over polarization) be a sum
+            # or an average?
+            power_beam[i] = np.sum(self.primary_beam.data_array[0,0,:,i], axis=0)
+            power_beam_int[i] = np.sum(power_beam[i]) # Integral of power beam
+            power_beam_Cl.append(hp.anafast(power_beam[i])) # Angular power spectra
+        power_beam_int *= 4.*np.pi / self.beam_model_npix
+        two_ell_plus_one = 2 * np.arange(3*int(np.sqrt(self.beam_model_npix/12))) + 1
+        sq_beam_integral = np.einsum('i,ji', two_ell_plus_one, power_beam_Cl)
+
+        lower_freq = self.freqs[0]
+        upper_freq = self.freqs[-1]
+        integration_freqs = np.linspace(lower_freq,upper_freq,num_steps)
+        integration_freqs_MHz = integration_freqs / 10**6 # The GP fitting below is more stable in MHz
+        redshifts = conversion.f2z(integration_freqs).flatten()
+        X2Y = np.array(map(conversion.X2Y,redshifts))
+
+        # Use a Gaussian Process to interpolate the frequency-dependent quantities
+        # derived from the beam model to the same frequency grid as the power spectrum
+        # estimation
+        beam_model_freqs_MHz = self.beam_freqs.T/(1.*10**6)
+        gp = GaussianProcessRegressor()
+        gp.fit(beam_model_freqs_MHz, sq_beam_integral/power_beam_int**2)
+        dOpp_over_Op2 = gp.predict(np.atleast_2d(integration_freqs_MHz).T)
+
+        # Get B_pp = \int dnu taper^2 and Bp = \int dnu
+        if taper == 'none':
+            dBpp_over_BpSq = np.ones_like(integration_freqs)
+        else:
+            dBpp_over_BpSq = aipy.dsp.gen_window(self.Nfreqs,'blackman-harris')**2
```

Comment:
This doesn't support user-defined window functions. It can only do 'none' or 'blackman-harris'.